### ISSUE TITLE
Colour support and winpty redirection broken

### ISSUE THREAD

**Issue by perlun** [perlang-org/perlang#285]
This is a followup to https://github.com/perlang-org/perlang/pull/284#discussion_r810635389.

## Background

The installation experience on Windows has historically been a bit lacking; see #107 for more details. I fixed this in #278, or so I thought.

After adding ANSI colors in #284, I tested the new version on Windows - only to realize that ANSI colors _did not work_ as expected:

![image](https://user-images.githubusercontent.com/630613/154907225-ac559f2f-f0c3-49c6-aadb-8e27c4ded8a5.png)

I also found out that in Git Bash (i.e. mingw-based `bash` on Windows), the fix I did in #278 _does not work_. Do note how the ANSI colors work in this case, though:

![image](https://user-images.githubusercontent.com/630613/154907528-85623b79-e35e-438e-9479-fc60ad494421.png)

This issue exists to dig further into this and try to understand why this happens, and find a mitigation.

**Comment by perlun** [comment]
Interestingly enough, when building a Debug build in Rider (on Windows), I get this:

![image](https://user-images.githubusercontent.com/630613/154908250-e37994d7-72d6-44c9-adf8-6b6cfe30ecc7.png)

In other words:

1. Debug builds, running on "Any CPU" seems to properly detect input being redirected. I've tested a Release build (built from Rider) and it behaves exactly the same.
2. When executed via `winpty`, ANSI colours do not seem to work. (Could there be a workaround to this? 🤔)

Either way, we need to figure out why these release-oriented builds work differently than the locally built Debug/Release builds.

**Comment by perlun** [comment]
> 2\. When executed via `winpty`, ANSI colours do not seem to work. (Could there be a workaround to this? 🤔)

This could be a variation of https://github.com/rprichard/winpty/issues/140 we're seeing. I'll ask a question there to see if anyone has any ideas.

**Comment by perlun** [comment]
Hmm, interesting. I was using a slightly older version Git for Windows, where the provided MSYS2 version was too old (`3.0.7-338.x86_64`) to support the more modern [ConPTY mode](https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/) introduced in Windows 10. @mintty helped me debug this in https://github.com/rprichard/winpty/issues/140#issuecomment-1046973609 (thanks 🙏).

After upgrading to a more recent version of Git for Windows, it worked flawlessly _with the existing Perlang version_. No changes were required (but we can probably drop the WinPTY re-execution call now since there's little need for it anymore):

![image](https://user-images.githubusercontent.com/630613/155009532-a3ee816e-4a6c-4d7f-9ab8-0e010581d05f.png)

### Important things to notice

ConPTY support is not enabled in the "Bash Git" included in Git for Windows by default. It must be enabled either by setting the `MSYS` env variable to `enable_pcon` before launching Git Bash, or by checking this setting during installation:

![image](https://user-images.githubusercontent.com/630613/155009665-952e756a-16aa-4f43-bab6-b76ce1372421.png)

**Comment by perlun** [comment]
For some odd reason, colours still doesn't work entirely correct when Git Bash is executed in the new, cool [Windows Terminal](https://github.com/microsoft/terminal). I think I'll leave this as a "known glitch" for now. 🤷‍♂️ 

![image](https://user-images.githubusercontent.com/630613/155010545-6ca52166-6477-4b38-9f29-983ce4a8c54e.png)

**Comment by perlun** [comment]
> For some odd reason, colours still doesn't work entirely correct when Git Bash is executed in the new, cool [Windows Terminal](https://github.com/microsoft/terminal).

Could be some variation of https://github.com/git-for-windows/git/issues/2483, will ask in that repo to see if anyone has any ideas. 🤔

**Comment by perlun** [comment]
No response yet. I'll leave this as a "known defect" in 0.1.0 for now, bumping this to 0.2.0. It would be nice to at least have a workaround for `bash` on Windows (MSYS2 or MINGW64) for this.

**Comment by perlun** [comment]
Still an open issue. I don't think this should block 0.3.0 though, so bumping once again.

**Comment by perlun** [comment]
> For some odd reason, colours still doesn't work entirely correct when Git Bash is executed in the new, cool [Windows Terminal](https://github.com/microsoft/terminal). I think I'll leave this as a "known glitch" for now. 🤷‍♂️

Interestingly enough, with the changes coming in #389, _this is no longer the case_ - colours _do_ work in the Windows Terminal:

![image](https://user-images.githubusercontent.com/630613/233724572-d934ab34-c0d1-4d50-881d-5051ddc5cd08.png)

...however, digging deeper, I tested with another build (893c5be), and it works equally well. 🤔 

I tested with 0.3.0 (i.e. e8c28e2) and it also worked equally well. 🤷 I guess we can consider this fixed then. Since it's unclear whether this is really:

- a fix on our side, _or_
- a fix in Windows Terminal/conpty, _or_
- a configuration change in the way I'm running Git Bash now vs previously

...I think we'll leave this out of the release notes for now. Regardless, closing issue.
